noir-projects/aztec-nr/aztec/src/macros/notes/mod.nr

```diff
@@ -28,18 +28,59 @@ comptime fn get_next_note_type_id() -> Field {
     note_type_id
 }
 
-/// Generates default `NoteInterface` implementation for a given note struct `s` and returns it as quote along with
-/// the length of the packed note.
+/// Generates a quote that implements `Packable` for a given note struct `s`.
+/// If the note struct already implements `Packable`, we return an empty quote.
+comptime fn derive_packable_if_not_implemented_and_get_len(s: StructDefinition) -> (Quoted, u32) {
+    // We try to get the packed length of the note struct. If it does not implement `Packable`, we get Option::none()
+    let packed_len_typ = std::meta::typ::fresh_type_variable();
+    // We don't care about the result of the implements check. We just want the get the packed length.
+    let _ = s.as_type().implements(
+        quote { crate::protocol_types::traits::Packable<$packed_len_typ> }.as_trait_constraint(),
+    );
+    let maybe_packed_length = packed_len_typ.as_constant();
+
+    if maybe_packed_length.is_some() {
+        // We got some packed length meaning that the note struct implements `Packable`. For this reason we return
+        // an empty quote for the implementation and the packed length.
+        (quote {}, maybe_packed_length.unwrap())
+    } else {
+        // We didn't manage to get the packed length which means the note struct doesn't implement `Packable`
+        // so we derive it and return it along with the packed length.
+        derive_packable_and_get_packed_len(s)
+    }
+}
+
+/// Generates default `NoteInterface` implementation for a given note struct `s` and returns it as a quote.
 ///
-/// impl NoteInterface<N> for NoteStruct {
-///     fn pack_content(self) -> [Field; N] {
+/// impl NoteInterface for NoteStruct {
+///     fn get_note_type_id() -> Field {
 ///         ...
 ///     }
 ///
-///     fn unpack_content(packed_content: [Field; N]) -> Self {
+///     fn compute_note_hash(self, storage_slot: Field) -> Field {
 ///         ...
 ///     }
+/// }
+comptime fn generate_note_interface(s: StructDefinition, note_type_id: Field) -> Quoted {
```

Comment:
I separated original `generate_note_interface` to `generate_note_interface_for_partial_note` and `generate_note_interface` because for normal notes we can use poseidon2 which is much cheaper. It also makes it clearer what is partial notes related complexity and what is not.